main-command/src/main/scala/sbt/internal/classpath/ClassLoaderCache.scala

```diff
@@ -0,0 +1,217 @@
+  override val commonParent: ClassLoader = configuration.provider.scalaProvider.launcher.topLoader
+  private val scalaProviderKey = configuration.provider.scalaProvider match {
+    case p: ExtendedScalaProvider =>
+      p.jars.find(_.getName == "scala-library.jar").map { l =>
+        new Key((l -> IO.getModifiedTimeOrZero(l)) :: Nil, p.launcher.topLoader) {
+          override def toClassLoader: ClassLoader = p.loaderLibraryOnly
+        }
+      }
+    case _ => None
+  }
+  private class Key(val fileStamps: Seq[(File, Long)], val parent: ClassLoader) {
+    def this(files: List[File]) =
+      this(files.map(f => f -> IO.getModifiedTimeOrZero(f)), commonParent)
+    lazy val files: Seq[File] = fileStamps.map(_._1)
+    lazy val maxStamp: Long = fileStamps.maxBy(_._2)._2
+    class CachedClassLoader
+        extends URLClassLoader(fileStamps.map(_._1.toURI.toURL).toArray, parent) {
+      override def toString: String =
+        s"CachedClassloader {\n  parent: $parent\n  urls:\n" + getURLs.mkString("    ", "\n", "\n}")
+    }
+    def toClassLoader: ClassLoader = new CachedClassLoader
+    override def equals(o: Any): Boolean = o match {
+      case that: Key => this.fileStamps == that.fileStamps && this.parent == that.parent
+    }
+    override def hashCode(): Int = (fileStamps.hashCode * 31) ^ parent.hashCode
+    override def toString: String = s"Key(${fileStamps mkString ","}, $parent)"
+  }
+  private[this] val delegate =
+    new java.util.concurrent.ConcurrentHashMap[Key, Reference[ClassLoader]]()
+  private[this] val referenceQueue = new ReferenceQueue[ClassLoader]
+  private[this] val metaspaceIsLimited =
+    ManagementFactory.getMemoryPoolMXBeans.asScala
+      .exists(b => (b.getName == "Metaspace") && (b.getUsage.getMax > 0))
+  private[this] class CleanupThread(private[this] val id: Int)
+      extends Thread(s"classloader-cache-cleanup-$id") {
+    setDaemon(true)
+    start()
+    private def cleanup(): Unit = {
+      val toClose = lock.synchronized(delegate.asScala.groupBy(_._1.files.toSet).flatMap {
+        case (_, pairs) if pairs.size > 1 =>
+          val max = pairs.maxBy(_._1.maxStamp)._1
+          pairs.filterNot(_._1 == max).flatMap {
+            case (k, v) =>
+              delegate.remove(k)
+              Option(v.get)
+          }
+        case _ => Nil
+      })
+      toClose.foreach(close)
+    }
+    @tailrec
+    override final def run(): Unit = {
+      val stop = try {
+        referenceQueue.remove(1000) match {
+          case ClassLoaderReference(key, classLoader) =>
+            close(classLoader)
+            delegate.remove(key)
+          case _ =>
+        }
+        cleanup()
+        false
+      } catch {
+        case _: InterruptedException => true
+      }
+      if (!stop) run()
+    }
+  }
+  private[this] val mkReference: (Key, ClassLoader) => Reference[ClassLoader] =
+    if (metaspaceIsLimited)(_, cl) => new SoftReference(cl, referenceQueue)
```

Comment:
I can't think of other places in sbt where we are doing something special based on the memory parameter like this.
It might be good to leave some comment here explaining the current tradeoff.

Reply:
Fair enough. It is mentioned in the commit message but this probably deserves higher visibility.

Reply:
https://github.com/sbt/sbt/pull/4729/commits/61bfba0dedd77fa12d7deebc6e6b69c21f18219f